torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -1036,45 +1029,84 @@ class ShapePropagator {
+    //   scalar type    : dtype if specified, else preserved
+    //   device         : preserved
+    //   tensor inputs  : 1
+    //   tensor outputs : 1
+    // Additionally:
+    //   - First input should be the only tensor input
+    static const register_formula_for reduce_ops_with_opt_dtype {
+      {
+        "aten::mean(Tensor self, *, int? dtype) -> Tensor"
+      },
+      [](Node* node) -> type_vec_t {
+        at::optional<IValue> maybe_dtype_option = node->get(attr::dtype);
+        if (auto type = node->input(0)->type()->cast<DimensionedTensorType>()) {
+          auto ret = type->withDim(0);
+          if(maybe_dtype_option && !maybe_dtype_option->isNone()) {
+            return {ret->toScalarType(maybe_dtype_option->toScalarType())};
+          } else {
+            return {ret};
+          }
+        }
+        return {};
+      }};
+
+    // Requirements:
+    //   dims           : 0
+    //   scalar type    : dtype if specified, else preserved if floating point, otherwise long/int64
     //   device         : preserved
     //   tensor inputs  : 1
     //   tensor outputs : 1
     // Additionally:
     //   - First input should be the only tensor input
-    static const register_formula_for all_reduce_ops_with_integer_upcast{
+    static const register_formula_for all_reduce_ops_with_integer_upcast_and_dtype{
         {
-            "aten::sum(Tensor self) -> Tensor",
-            "aten::prod(Tensor self) -> Tensor",
+            "aten::sum(Tensor self, *, int? dtype) -> Tensor",
+            "aten::prod(Tensor self, *, int? dtype) -> Tensor",
         },
         [](Node* node) -> type_vec_t {
           if (auto type =
                   node->input(0)->type()->cast<DimensionedTensorType>()) {
+            at::optional<IValue> maybe_dtype_option = node->get(attr::dtype);
+            if( maybe_dtype_option && ! maybe_dtype_option->isNone() ) {
+              return {type->withDim(0)->toScalarType(maybe_dtype_option->toScalarType())};
+            }
             return {at::isFloatingType(type->scalarType())
                         ? type->withDim(0)
                         : type->withDim(0)->toScalarType(at::kLong)};
           }
           return {};
         }};
 
-    static const auto multidim_reduce_with_postprocess =
+    static const auto reduce_op_handler =
+      [](Node* node,
+        int64_t num_reduced_dim = 0,
+        bool upcast_integer = false,
+        c10::optional<IValue> opt_dtype = c10::nullopt) -> type_vec_t {
+          if(auto type = node->input(0)->type()->cast<DimensionedTensorType>()) {
+            if( opt_dtype && ! opt_dtype->isNone() ) {
+              type = type->toScalarType(opt_dtype->toScalarType());
+            } else if(upcast_integer && !at::isFloatingType(type->scalarType())) {
+              type = type->toScalarType(at::kLong);
+            }
+            if (type->dim() >= num_reduced_dim && num_reduced_dim > 0) {
```

Comment:
equivalent previous code was `if (type->dim() > num_reduced_dim)` which seems to have been incorrect / a bug. Wasn't allowing for reductions from 1 to 0 dim.